src/test/app/TrustAndBalance_test.cpp

```diff
@@ -0,0 +1,512 @@
+                (with_rate ? "With " : "Without ") + " Xfer Fee, ");
+        using namespace test::jtx;
+
+        Env env {*this};
+        Account gw {"gateway"};
+        Account amazon {"amazon"};
+        Account alice {"alice"};
+        Account bob {"bob"};
+        Account carol {"carol"};
+
+        env.fund (XRP(10000), gw, amazon, alice, bob, carol);
+        env.close();
+
+        env (trust (amazon, gw["USD"](2000)));
+        env (trust (bob, alice["USD"](600)));
+        env (trust (bob, gw["USD"](1000)));
+        env (trust (carol, alice["USD"](700)));
+        env (trust (carol, gw["USD"](1000)));
+
+        if (with_rate)
+            env (rate (gw, 1.1));
+
+        env (pay (gw, bob, bob["USD"](100)));
+        env (pay (gw, carol, carol["USD"](100)));
+        env.close();
+
+        // alice pays amazon via multiple paths
+        if (with_rate)
+            env (pay (alice, amazon, gw["USD"](150)),
+                sendmax(alice["USD"](200)),
+                test::jtx::path(bob),
+                test::jtx::path(carol));
+        else
+            env (pay (alice, amazon, gw["USD"](150)),
+                test::jtx::path(bob),
+                test::jtx::path(carol));
+
+        if (with_rate)
+        {
+            // 65.00000000000001 is correct.
+            // This is result of limited precision.
+            env.require (balance (
+                alice,
+                STAmount(
+                    carol["USD"].issue(),
+                    6500000000000001ull,
+                    -14,
+                    false,
+                    true,
+                    STAmount::unchecked{})));
+            env.require (balance (carol, gw["USD"](35)));
+        }
+        else
+        {
+            env.require (balance (alice, carol["USD"](-50)));
+            env.require (balance (carol, gw["USD"](50)));
+        }
+        env.require (balance (alice, bob["USD"](-100)));
+        env.require (balance (amazon, gw["USD"](150)));
+        env.require (balance (bob, gw["USD"](0)));
+    }
+
+    void
+    testInvoiceID ()
+    {
+        testcase ("Set Invoice ID on Payment");
+        using namespace test::jtx;
+
+        Env env {*this};
+        Account alice {"alice"};
+        auto wsc = test::makeWSClient(env.app().config());
+
+        env.fund (XRP(10000), alice);
+        env.close();
+
+        Json::Value jvs;
+        jvs[jss::accounts] = Json::arrayValue;
+        jvs[jss::accounts].append(env.master.human());
+        jvs[jss::streams] = Json::arrayValue;
+        jvs[jss::streams].append("transactions");
+        BEAST_EXPECT(wsc->invoke("subscribe", jvs)[jss::status] == "success");
+
+        Json::Value jv;
+        auto tx = env.jt (
+            pay (env.master, alice, XRP(10000)),
+            json(sfInvoiceID.fieldName, "DEADBEEF"));
+        jv[jss::tx_blob] = strHex (tx.stx->getSerializer().slice());
+        auto jrr = env.rpc ("json", "submit", to_string(jv)) [jss::result];
```

Comment:
You can use the websocket connection you created.
`auto jrr = wsc->invoke("submit", jv);`